ampd/src/grpc/ampd.rs

```diff
@@ -1,37 +1,402 @@
+            }),
+        }
+    }
+}
+
+impl proto::SubscribeRequest {
+    fn matches(&self, event: &Event) -> bool {
+        self.event_filter
+            .as_ref()
+            .map(|event_filter| match event {
+                Event::BlockBegin(_) | Event::BlockEnd(_) => event_filter.include_block_begin_end,
+                Event::Abci {
+                    event_type,
+                    attributes,
+                } => {
+                    event_filter.filters.is_empty()
+                        || event_filter.filters.iter().any(|filter| {
+                            filter.event_type == *event_type
+                                && filter.event_attributes.iter().all(|(key, value)| {
+                                    attributes.get(key).map(|v| v == value).unwrap_or_default()
+                                })
+                        })
+                }
+            })
+            .unwrap_or(true)
+    }
+}
+
+pub struct Server<S, B>
+where
+    S: EventSub,
+    B: BroadcasterClient,
+{
+    event_subscriber: S,
+    broadcaster: B,
+}
 
-impl Ampd {
-    pub fn new() -> Self {
-        Self {}
+impl<S, B> Server<S, B>
+where
+    S: EventSub,
+    B: BroadcasterClient,
+{
+    pub fn new(event_subscriber: S, broadcaster: B) -> Self {
+        Self {
+            event_subscriber,
+            broadcaster,
+        }
     }
 }
 
 #[async_trait]
-impl proto::ampd_server::Ampd for Ampd {
+impl<S, B> proto::ampd_server::Ampd for Server<S, B>
+where
+    S: EventSub + Send + Sync + 'static,
+    B: BroadcasterClient + Send + Sync + 'static,
+{
     type SubscribeStream =
         Pin<Box<dyn Stream<Item = Result<proto::SubscribeResponse, Status>> + Send + 'static>>;
 
     async fn subscribe(
         &self,
-        _req: Request<proto::SubscribeRequest>,
+        req: Request<proto::SubscribeRequest>,
     ) -> Result<Response<Self::SubscribeStream>, Status> {
-        todo!()
+        let req = req.into_inner();
+        let stream = self
+            .event_subscriber
+            .subscribe()
+            .filter(move |event| {
+                future::ready(
+                    event
+                        .as_ref()
+                        .map(|event| req.matches(event))
+                        .unwrap_or(true),
+                )
+            })
+            .map_ok(Into::into)
+            .map_ok(|event| proto::SubscribeResponse { event: Some(event) })
+            .map_err(|err| Status::internal(format!("failed to subscribe to events: {}", err)));
+
+        Ok(Response::new(Box::pin(stream)))
     }
 
     async fn broadcast(
         &self,
-        _req: Request<proto::BroadcastRequest>,
+        req: Request<proto::BroadcastRequest>,
     ) -> Result<Response<proto::BroadcastResponse>, Status> {
-        todo!()
+        let req = req.into_inner();
+        let msg = req.msg.ok_or(Status::invalid_argument("missing msg"))?;
+
+        // TODO: validate msg by estimating its gas cost before broadcasting?
```

Comment:
is the broadcaster not doing that?

Reply:
It does, but not at the point when `BroadcasterClient` pushes it through. Esstially, the `BroadcasterClient` wouldn't get an error for that even if it happens.